main/src/main/scala/sbt/internal/server/BuildServerProtocol.scala

```diff
@@ -840,15 +846,29 @@ object BuildServerProtocol {
     }
   }
 
-  private def scalaTestClassesTask: Initialize[Task[ScalaTestClassesItem]] = Def.task {
-    val testClasses = Keys.definedTests.?.value
-      .getOrElse(Seq.empty)
-      .map(_.name)
-      .toVector
-    ScalaTestClassesItem(
-      bspTargetIdentifier.value,
-      testClasses
-    )
+  private def scalaTestClassesTask: Initialize[Task[Seq[ScalaTestClassesItem]]] = Def.task {
+    Keys.definedTests.?.value match {
+      case None => Vector.empty
+      case Some(defined) =>
+        val fingerprints = Keys.loadedTestFrameworks.value.values.flatMap { framework =>
+          framework.fingerprints().map(fingerprint => (fingerprint, framework))
+        }.toMap
+        defined
+          .map(
+            definition =>
+              TestDefinitionWithFramework(definition, fingerprints.get(definition.fingerprint))
+          )
+          .groupBy(_.framework)
```

Comment:
I don't think we need this `TestDefinitionWithFramework` anymore.

```suggestion
        defined
          .groupBy(def => fingerprints.get(definition.fingerprint))
```